### PR TITLE
Introduce basic support for notification-triggers in device.launchApp()

### PR DESCRIPTION
**Pull request by d4vidi**
- [ ] This is a small change 
- [x] This change has been discussed in issue #1192 and the solution has been agreed upon with maintainers.

---

**Description:**
Partially implements #1192 by adding support for the core use case, as conveyed by this API:
```js
    const userNotification = {
      payload: {
        userData1: 'userDataValue1',
        userData2: 'userDataValue2',
      },
    };
    await device.launchApp({ newInstance: true, userNotification });
```

Gaps to fill in after merge:
- [ ] Add support for handling notification when app is in background (i.e. move to foreground with notification data)
- [ ] Add support for `sendUserNotification()` API

### CODE REVIEW

File: detox/android/detox/src/main/java/com/wix/detox/LaunchArgs.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import android.os.Bundle;
+import android.util.Base64;
+
+import java.util.Arrays;
+import java.util.List;
+
+import androidx.test.platform.app.InstrumentationRegistry;
+
+class LaunchArgs {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/LaunchArgs.java:11]: Extracted from `Detox`, with the addition of the notification-path arg

File: detox/android/detox/src/main/java/com/wix/detox/DetoxIdlePolicyConfig.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.concurrent.TimeUnit;
+
+import androidx.test.espresso.IdlingPolicies;
+
+/**
+ * Specification of values to use for Espresso's {@link IdlingPolicies} timeouts.
+ * <br/>Overrides Espresso's defaults as they tend to be too short (e.g. when running on heavy-load app
+ * on suboptimal CI machines).
+ */
+public class DetoxIdlePolicyConfig {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxIdlePolicyConfig.java:12]: Extracted from `Detox`, alongside the new `apply()` method so as to move the responsibility where it belongs

File: detox/android/detox/src/main/java/com/wix/detox/LaunchIntentsFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package com.wix.detox;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+
+class LaunchIntentsFactory {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/LaunchIntentsFactory.java:9]: Extracted from `Detox` (thank god), with `intentWithNotificationData` as the new method to handle this use case specificially

File: detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import org.assertj.core.api.Assertions.assertThat
+import org.json.JSONArray
+import org.json.JSONObject
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import java.lang.IllegalArgumentException
+import kotlin.test.assertFailsWith
+
+@RunWith(RobolectricTestRunner::class)
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt:12]: Can't test code that uses `Bundle` withouth Robolectric

File: detox/src/devices/drivers/DeviceDriverBase.js [hunk 5]
```diff
@@ -108,7 +108,10 @@ class DeviceDriverBase {
   }
 
   createPayloadFile(notification) {
-
+    const fs = require('fs');
```
Comment (d4vidi) [inline detox/src/devices/drivers/DeviceDriverBase.js:111]: moved from IosDriver

File: detox/src/devices/drivers/android/tools/AppInstallHelper.js [hunk 6]
```diff
@@ -1,35 +1,26 @@
-const EMU_TEMP_PATH = '/data/local/tmp';
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/AppInstallHelper.js:1]: Extracted file-transferring logic onto a dedicated class (`TempFileXfer`)

File: detox/test/src/app.js [hunk 7]
```diff
@@ -58,93 +65,81 @@ class example extends Component {
...
-
-    }
-
-    else if (this.state.url) {
+    if (this.state.url) {
       console.log("App@render: rendering a URL:", this.state.url);
       return this.renderText(this.state.url);
     }
 
-    if (!this.state.screen) {
```
Comment (d4vidi) [inline detox/test/src/app.js:87]: ugh... indent changes...

File: detox/android/detox/src/main/java/com/wix/detox/Detox.java [hunk 8]
```diff
@@ -91,6 +87,11 @@
         public Integer masterTimeoutSec = 120;
         /** Directly binds to {@link IdlingPolicies#setIdlingResourceTimeout(long, TimeUnit)}. Applied in seconds. */
         public Integer idleResourceTimeoutSec = 60;
+
+        void apply() {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/Detox.java:91]: Moved to the right context -- replaces `Detox. applyIdlePolicyConfig()`

File: detox/android/detox/src/main/java/com/wix/detox/LaunchIntentsFactory.kt [hunk 9]
```diff
@@ -0,0 +1,102 @@
+package com.wix.detox
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+
+internal class LaunchIntentsFactory {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/LaunchIntentsFactory.kt:8]: Extracted from `Detox`, alongside the additional `intentWithNotificationData` method that is used in the notification-data handling flow

File: detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+private fun aJsonWithALongInt() = JSONObject().put("longint", 1234L)
+private fun aJsonWithADouble() = JSONObject().put("double", 1234.56)
+private fun aJsonWithAChildObject(): JSONObject {
+    val child = JSONObject("""{
+        "flag": true,
+        "test": "777",
+        "integer": 1234,
+        "longint": 2222222222,
+        "double": 1234.56
+    }""".trimIndent())
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt:118]: @valentynberehovyi ez way to create jsons using kotlin -- I recall you had a similar case in some of your unit tests

File: detox/src/devices/drivers/android/tools/AppInstallHelper.js [hunk 11]
```diff
@@ -1,35 +1,26 @@
...
-const EMU_TEMP_INSTALL_PATH = `${EMU_TEMP_PATH}/detox`;
-
 class AppInstallHelper {
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/AppInstallHelper.js:1]: Extracted the contained logic of transferring files onto the device to a dedicated class (`FileXfer`)

File: detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt (hunk 10 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/test/java/com/wix/detox/common/JsonConverterTest.kt:118]: Very nice, @d4vidi !
I'll use this way then.

File: detox/src/devices/drivers/DeviceDriverBase.js (hunk 5 above)
Comment (noomorph) [reply detox/src/devices/drivers/DeviceDriverBase.js:111]: Is requiring `fs` so lately caused by unit tests?

Comment (d4vidi) [reply detox/src/devices/drivers/DeviceDriverBase.js:111]: no - just residue from an intermediate refactor state. thanks for pointing out!

**Pull request merged**